x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -132,6 +135,50 @@ private void getRoles(User user, Authentication authentication, ActionListener<R
         rolesStore.getRoles(user, authentication, listener);
     }
 
+    @Override
+    public boolean isChildActionAuthorizedByParent(RequestInfo requestInfo, String parentAction, IndicesAccessControl parentAccessControl) {
+        final String childAction = requestInfo.getAction();
+        if (IndexPrivilege.CREATE_INDEX_MATCHER.test(childAction) || childAction.equals(TransportShardBulkAction.ACTION_NAME)) {
+            // These need special handling, so don't short circuit
+            return false;
+        }
+        if (isScrollRelatedAction(childAction) || isAsyncRelatedAction(childAction)) {
+            // These need special handling, so don't short circuit
+            return false;
+        }
+        if (childAction.startsWith(parentAction) == false) {
+            // Parent action is not a true parent
+            // We want to treat shard level actions (those that append '[s]' and/or '[p]' & '[r]')
+            // or similar (e.g. search phases) as children, but not every action that is triggered
+            // within another action should be authorized this way
+            return false;
+        }
+        final IndicesRequest indicesRequest;
+        if (requestInfo.getRequest() instanceof IndicesRequest) {
+            indicesRequest = (IndicesRequest) requestInfo.getRequest();
+        } else {
+            // Can only handle indices request here
+            return false;
+        }
+
+        final String[] indices = indicesRequest.indices();
+        if (indices == null || indices.length == 0) {
+            // No indices to check
+            return false;
+        }
+
+
+        return Arrays.stream(indices).allMatch(idx -> {
+            if (Regex.isSimpleMatchPattern(idx)) {
+                // The request contains a wildcard
+                return false;
+            }
+            IndicesAccessControl.IndexAccessControl iac = parentAccessControl.getIndexPermissions(idx);
+            // The parent context has already successfully authorized access to this index (by name)
+            return iac != null && iac.isGranted();
```

Comment:
I cannot think of a real issue for it. But would like to write down my thinking here in case others could find out more about it.

Is it possible for an index of child action having different `IndexAccessControl` than the parent? So even when `iac != null && iac.isGranted()` returns true, `iac.getDocumentPermissions` may not be the same for them? A terms lookup can have this effect. But action of fetching document from another index is not a child action of the original search action (name wise). So it is not really an issue. But there maybe other things that I didn't think of?

Reply:
In theory it can have less restrictions, but not more.
I'm pretty sure that would also break other things.

You could have:
```
"indices": [ 
 {
   "names": [ "*" ], "privileges": [ "indices:data/read/search" ], "query": "something"  
 },
 {
   "names": [ "*" ], "privileges": [ "indices:data/read/search[" ], "query": null
 }
]
```
which would mean searching on an index would apply DLS but all the phases and the shard level searches would not have DLS.
I don't think we claim to support that - we certainly don't test for it, nor should we.

The reverse:
```
"indices": [ 
 {
   "names": [ "*" ], "privileges": [ "indices:data/read/search" ], "query": null 
 },
 {
   "names": [ "*" ], "privileges": [ "indices:data/read/search[" ], "query": "something"
 }
]
```
would not be relevant because the first `Group` would apply to all the phases (per `Privilege.actionToPattern`) so and the `null` DLS query would take precedence.